File: js/balanceParens.js
balanceParens = (str) => {
    let stack = []
    let remove = []
    for (let i = 0; i < str.length; i++) {
        if(str[i] === '('){
            stack.push(i)
        } else if(str[i] === ')' && stack.length > 0){
            stack.pop()
        } else if(str[i] === '(' || str[i] === ')' && stack.length === 0){
            remove.push(i)
        }
    }

    if(stack.length > 0){
        for(let num in stack){
            remove.push(stack[num])
        }
    }

    let result = []
    for(let i = 0; i < str.length; i++){
        if(remove.indexOf(i) !== -1){
            continue
        } else {
            result.push(str[i])
        }
    }
    return result.join('')
}

module.exports = { balanceParens }

File: python/balance_parens.py
def balance_parens(str):
    stack = []
    remove = [] #keep track of indexes
    for i, x in enumerate(str):
        if x == '(':
            stack.append(i)
        elif x == ')' and len(stack) > 0:
            stack.pop()
        elif (x == '(' or x == ')') and len(stack) == 0:
            remove.append(i)
    
    if len(stack) > 0:
        for num in stack:
            remove.append(num)
    
    result = []
    for i, x in enumerate(str):
        if i in remove:
            continue
        else:
            result.append(x)
            
    return ''.join(result)
        
